inc/class-job.php

```diff
@@ -186,39 +194,184 @@ public static function get_by_site( $site, $include_completed = false, $include_
 			return new WP_Error( 'cavalcade.job.invalid_site_id' );
 		}
 
-		if ( ! $include_completed && ! $include_failed ) {
-			$results = wp_cache_get( 'jobs', 'cavalcade-jobs' );
+		$statuses = [ 'waiting', 'running' ];
+		if ( $include_completed ) {
+			$statuses[] = 'completed';
+		}
+		if ( $include_failed ) {
+			$statuses[] = 'failed';
 		}
 
-		if ( isset( $results ) && ! $results ) {
-			$statuses = [ 'waiting', 'running' ];
-			if ( $include_completed ) {
-				$statuses[] = 'completed';
-			}
-			if ( $include_failed ) {
-				$statuses[] = 'failed';
-			}
+		$results = static::get_jobs_by_query(
+			[
+				'site' => $site,
+				'args' => null,
+				'statuses' => $statuses,
+				'limit' => 0,
+				'__raw' => true,
+			]
+		);
 
-			// Find all scheduled events for this site
-			$table = static::get_table();
+		if ( empty( $results ) ) {
+			return [];
+		}
 
-			$sql = "SELECT * FROM `{$table}` WHERE site = %d";
-			$sql .= ' AND status IN(' . implode( ',', array_fill( 0, count( $statuses ), '%s' ) ) . ')';
-			$query = $wpdb->prepare( $sql, array_merge( [ $site ], $statuses ) );
-			$results = $wpdb->get_results( $query );
+		return static::to_instances( $results );
+	}
 
-			if ( ! $include_completed && ! $include_failed ) {
-				wp_cache_set( 'jobs', $results, 'cavalcade-jobs' );
-			}
+	/**
+	 * Query jobs database.
+	 *
+	 * Returns an array of Job instances for the current site based
+	 * on the paramaters.
+	 *
+	 * @todo: allow searching within time window for duplicate events.
+	 *
+	 * @param array|\stdClass $args {
+	 *     @param string          $hook      Jobs hook to return. Optional.
+	 *     @param int|string|null $timestamp Timestamp to search for. Optional.
+	 *                                       String shortcuts `future`: > NOW(); `past`: <= NOW()
+	 *     @param array           $args      Cron job arguments.
+	 *     @param int|object      $site      Site to query. Default current site.
+	 *     @param array           $statuses  Job statuses to query. Default to waiting and running.
+	 *     @param int             $limit     Max number of jobs to return. Default 1.
+	 *     @param string          $order     ASC or DESC. Default ASC.
+	 * }
+	 * @return Job[]|WP_Error Jobs on success, error otherwise.
+	 */
+	public static function get_jobs_by_query( $args = [] ) {
+		global $wpdb;
+		$args = (array) $args;
+		$results = [];
+
+		$defaults = [
+			'timestamp' => null,
+			'hook' => null,
+			'args' => [],
+			'site' => get_current_blog_id(),
+			'statuses' => [ 'waiting', 'running' ],
+			'limit' => 1,
+			'order' => 'ASC',
+			'__raw' => false,
+		];
+
+		$args = wp_parse_args( $args, $defaults );
```

Comment:
Can we add a filter after this as well? Will allow implementing that max limit check in more opinionated code.